src/relay/ir/hash.cc

```diff
@@ -0,0 +1,322 @@
+
+    if (ref->derived_from<TypeNode>()) {
+      return TypeHash(Downcast<Type>(ref));
+    }
+    if (ref->derived_from<ExprNode>()) {
+      return ExprHash(Downcast<Expr>(ref));
+    }
+    return AttrHash(ref);
+  }
+
+  /*!
+   * Compute hash of the attributes.
+   * \param ref The attributes.
+   * \return the hash value
+   */
+  size_t AttrHash(const NodeRef& ref) {
+    if (!ref.defined()) { return ref.hash(); }
+    return AttrsHashHandler::Hash(ref);
+  }
+  /*!
+   * Compute hash of a Relay type.
+   * \param ref The type to hash.
+   * \param rhs The right hand operand.
+   * \return the hash value.
+   */
+  size_t TypeHash(const Type& type) {
+    if (!type.defined()) { return type.hash(); }
+    auto found = hash_map_.find(type);
+    if (found != hash_map_.end()) {
+      return found->second;
+    } else {
+      auto hash = this->VisitType(type);
+      hash_map_.insert({type, hash});
+      return hash;
+    }
+  }
+  /*!
+   * Compute the hash of an expression.
+   *
+   * \note We run graph structural equality checking when comparing two Exprs.
+   *   This means that AlphaEqualHandler can only be used once for each pair.
+   *   The equality checker checks data-flow equvalence of the Expr DAG.
+   *   This function also runs faster as it memomizes equal_map.
+   *
+   * \param expr The expression to hash.
+   * \return the hash value.
+   */
+  size_t ExprHash(const Expr& expr) {
+    if (!expr.defined()) return expr.hash();
+    auto found = hash_map_.find(expr);
+    if (found != hash_map_.end()) {
+      return found->second;
+    } else {
+      auto hash = this->VisitExpr(expr);
+      hash_map_.insert({expr, hash});
+      return hash;
+    }
+  }
+
+ protected:
+  /*!
+   * \brief Hash a DataType.
+   * \param dtype The dtype to hash.
+   * \return the hash value.
+   */
+  size_t DataTypeHash(const DataType& dtype) {
+     return std::hash<int>()(
+         static_cast<int>(dtype.code()) |
+         (static_cast<int>(dtype.bits()) << 8) |
+         (static_cast<int>(dtype.lanes()) << 16));
+  }
+
+  /*!
+   * \brief Check Equality of leaf node of the graph.
+   *  if map_free_var_ is set to true, try to map via equal node.
+   * \param lhs The left hand operand.
+   * \param rhs The right hand operand.
+   * \return the compare result.
+   */
+  size_t LeafNodeEqual(const NodeRef& lhs, const NodeRef& rhs) {
+    return 0;
+    // if (lhs.same_as(rhs)) return true;
+    // auto it = equal_map_.find(lhs);
+    // if (it != equal_map_.end()) {
+    //   return it->second.same_as(rhs);
+    // } else {
+    //   if (map_free_var_) {
+    //     if (lhs->type_index() != rhs->type_index()) return false;
+    //     equal_map_[lhs] = rhs;
+    //     return true;
+    //   } else {
+    //     return false;
+    //   }
+    // }
+  }
+
+  using AttrsHashHandler::VisitAttr_;
+  size_t VisitAttr_(const Variable* lhs) final {
+    return 0; // return LeafNodeEqual(GetRef<NodeRef>(lhs), other);
```

Comment:
@tqchen I don't fully understand this case.

Reply:
There are two possible cases in here:
- Variable is a var defined in say TypeParams, in this case, then we will need to put the hash value at the declaration site(likely you can put everything under the same hash_map_ and reuse it for both Attr and Expr
- Variable is a free variable, in that case, depending on the equality semantics(graph equal vs alpha equal)
   - In the case of alpha_equal, free variables do not map to each other, and hash by the pointer is a good choice
   - In the case of graph_equal, free variables might match each other, maybe we could hash by name_hint
   - By considering both cases, hash by name_hint might not be a bad choice